packages/iov-stream/src/concat.ts

```diff
@@ -0,0 +1,102 @@
+// tslint:disable:readonly-array
+import { Producer, Stream, Subscription } from "xstream";
+
+/**
+ * An implementation of concat that buffers all source stream events
+ *
+ * Marble diagram:
+ *
+ * ```text
+ * --1--2---3---4-|
+ * -a--b-c--d-|
+ * --------X---------Y---------Z-
+ *           concat
+ * --1--2---3---4-abcdXY-------Z-
+ * ```
+ *
+ * This is inspired by RxJS's concat as documented at http://rxmarbles.com/#concat and behaves
+ * differently than xstream's concat as discussed in https://github.com/staltz/xstream/issues/170.
+ *
+ */
+export function concat<T>(...streams: Array<Stream<T>>): Stream<T> {
+  const subscriptions = new Array<Subscription>();
+  const queues = new Array<T[]>(); // one queue per stream
+  const completedStreams = new Set<number>();
+  let activeStreamIndex = 0;
+
+  function reset(): void {
+    while (subscriptions.length > 0) {
+      const subscription = subscriptions.shift()!;
+      subscription.unsubscribe();
+    }
+
+    // tslint:disable-next-line:no-object-mutation
+    queues.length = 0;
+    completedStreams.clear();
+    activeStreamIndex = 0;
+  }
+
+  const producer: Producer<T> = {
+    start: listener => {
+      streams.forEach(_ => queues.push([]));
+
+      function emitAllQueuesEvents(streamIndex: number): void {
+        while (true) {
+          const element = queues[streamIndex].shift();
+          if (element === undefined) {
+            return;
+          }
+          listener.next(element);
+        }
+      }
+
+      function isDone(): boolean {
+        return activeStreamIndex >= streams.length;
+      }
+
+      if (isDone()) {
+        listener.complete();
+        return;
+      }
+
+      streams.forEach((stream, index) => {
```

Comment:
Okay, from inspection, I think the code is correct (wrt the test case I mentioned above).
You do subscribe to all streams at once and add them to queues to be delivered later